### ISSUE TITLE
Assertion `E->isNonOdrUse() != NOUR_Unevaluated && "should not emit an unevaluated operand"' with sizeof

### ISSUE THREAD

**Issue by efriedma-quic** [llvm/llvm-project#42425]
|  |  |
| --- | --- |
| Bugzilla Link | [43080](https://llvm.org/bz43080) |
| Version | unspecified |
| OS | Windows NT |
| CC | @DougGregor,@DimitryAndric,@JVApen,@miyuki,@riccibruno,@zygoloid |

## Extended Description 
Testcase (C++):

```cpp
int f(int i) { return sizeof i<i; }
```

clang-10: [...]/llvm/tools/clang/lib/CodeGen/CGExpr.cpp:2518: clang::CodeGen::LValue clang::CodeGen::CodeGenFunction::EmitDeclRefLValue(const clang::DeclRefExpr *): Assertion `E->isNonOdrUse() != NOUR_Unevaluated && "should not emit an unevaluated operand"' failed.

What I think is happening is that the parser sees "i<", so ParseCastExpression calls checkPotentialAngleBracket.  That tries to speculatively parse the second "i", to try to disambiguate a template-id.  It then backs out, but at that point we've already annotated the token with a DeclRefExpr.  That DeclRefExpr is incorrectly marked unevaluated because "isUnevaluatedContext()" is true when the expression is created.

We could probably teach mightBeIntendedToBeTemplateName so it knows that a local variable is never a template name... that avoids the problem here.  And I think we can always determine whether a expression is "considered to refer to a template" per [temp.names]?  But that makes diagnostics worse on some testcases. And I'm not sure checkPotentialAngleBracket is the only way to trigger this sort of issue.

Alternatively, we could try to solve the issue at the point we parse the annotated token.  Maybe we could throw away the expression?  Rewrite the expression it like we do typeid() expressions?  Something else?

The assertion is new in r363295, but the issue isn't, I think.  This particular formulation triggers a different assertion on older versions of clang.

**Comment by shafik** [comment]
I can't reproduce the crashes given here, please let me know if any of these are still valid problems.

**Comment by Endilll** [comment]
All three examples appear to be fixed in Clang 10:
https://cexplorer.testlabs.pro/z/8cYvKK
https://cexplorer.testlabs.pro/z/PevEdq
https://cexplorer.testlabs.pro/z/ahWqeE
